### PR TITLE
Illusion of Moonlight

### PR DESCRIPTION
**Pull request by Everade**
* **Addressed Issue(s)**: 
https://github.com/rathena/rathena/issues/6453

* **Server Mode**: 
Renewal

* **Description of Pull Request**: 
Implementation of Illusion of Moonlight (Walkthrough Conversion)

Credits to @crazy-arashi for the initial script release

### CODE REVIEW

File: npc/re/quests/quests_16_2_illusion.txt [hunk 1]
```diff
@@ -0,0 +1,1039 @@
...
+	questinfo QTYPE_QUEST,QMARK_YELLOW,"illusion_moonlight == 8";
+	end;
+}
+
+payon_in03,30,96,5	script	Jagyeom#ill_moon_pin	4_M_HUOLDARMY,{
+	if (illusion_moonlight < 2) {
+		mes "[ Jagyeom ]";
+		mes "Please observe proper manners inside the palace.";
+		close;
+	}
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:529]: The default text is at the end
```suggestion
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 2]
```diff
@@ -0,0 +1,1039 @@
...
+	if (illusion_moonlight == 2) {
+		npctalk "Muyeon... If you can't do it, no one else can!","",bc_self;
+		end;
+	}
+	if (illusion_moonlight < 5) {
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:534]:
```suggestion
	if (illusion_moonlight == 3 || illusion_moonlight == 4) {
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 3]
```diff
@@ -0,0 +1,1039 @@
...
+		npctalk "Let's listen for now.","",bc_self;
+		end;
+	}
+	if (illusion_moonlight == 5) {
+		mes "[ Jagyeom ]";
+		mes "We should grasp the situation first.","It'll be great if you can help us, Adventurer.";
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:540]:
```suggestion
	if (illusion_moonlight == 5 || illusion_moonlight == 8) {
		mes "[ Jagyeom ]";
		mes "We should grasp the situation first.";
		mes "It'll be great if you can help us, Adventurer.";
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 4]
```diff
@@ -0,0 +1,1039 @@
...
+	if (illusion_moonlight < 10) {
+		mes "[ Jagyeom ]";
+		mes "Please observe proper manners inside the palace.";
+		close;
+	}
+	mes "[ Jagyeom ]";
+	mes "The Moonlight Flower are recreating the past in their nightmares.","In a way, they're also victims.";
+	next;
+	mes "[ Jagyeom ]";
+	mes "I'll make sure no villagers gets close to the collapsed walls of reality and dreams.";
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:552]:
```suggestion
	if (illusion_moonlight == 9) {
		mes "[ Jagyeom ]";
		mes "The Moonlight Flower are recreating the past in their nightmares.";
		mes "In a way, they're also victims.";
		next;
		mes "[ Jagyeom ]";
		mes "I'll make sure no villagers gets close to the collapsed walls of reality and dreams.";
		close;
	}
	mes "[ Jagyeom ]";
	mes "Please observe proper manners inside the palace.";
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 5]
```diff
@@ -0,0 +1,1039 @@
...
+		mes "Then, do you mind if I tag along?","You don't have to worry about me. Junghee can keep me safe.";
+		next;
+		npctalk "D-Don't touch anything!","Junghee#ill_moon_pdun",bc_self;
+		mes "[ Young Scholar ]";
+		mes "Then... should I touch this?";
+		illusion_moonlight = 6;
+		completequest 7780;
+		setquest 7781;
+		close2;
+		specialeffect EF_SPINMOVE,AREA,"Young Scholar#ill_moon_pdun";
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:618]:
```suggestion
		setpcblock PCBLOCK_NPC, true;
		specialeffect EF_SPINMOVE,AREA,"Young Scholar#ill_moon_pdun";
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 6]
```diff
@@ -0,0 +1,1039 @@
...
+		npctalk "Arghhhhh-!","",bc_self;
+		cloakonnpc "Young Scholar#ill_moon_pdun",getcharid(0);
+		sleep2 500;
+		npctalk "Chunghae!","Junghee#ill_moon_pdun",bc_self;
+		sleep2 2000;
+		npctalk "That careless klutz!","Junghee#dun_dummy",bc_self;
+		sleep2 500;
+		specialeffect EF_SPINMOVE,AREA,"Junghee#ill_moon_pdun";
+		specialeffect EF_DQ9_CHARGE,AREA,"Junghee#ill_moon_pdun";
+		sleep2 350;
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:630]:
```suggestion
		sleep2 350;
		setpcblock PCBLOCK_NPC, false;
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 7]
```diff
@@ -0,0 +1,1039 @@
...
+	mes "[ Villager ]";
+	mes "We don't worship monsters.";
+	mes "We're not enthralled by monsters!";
+	next;
+	mes "[ Villager ]";
+	mes "Ah...";
+	mes "The light... The magic... Please stop that magic...";
+	close2;
+	initnpctimer;
+	disablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:951]:
```suggestion
	disablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 8]
```diff
@@ -0,0 +1,1039 @@
...
+	specialeffect EF_QUAKEBODY2;
+	npctalk "Ah... Ah... Aaahhh...";
+	areamonster "pay_d03_i",42,149,44,151,"Resentful Sohee",3762,3;
+	end;
+
+OnTimer60000:
+	stopnpctimer;
+	enablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:959]:
```suggestion
	enablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 9]
```diff
@@ -0,0 +1,1039 @@
...
+}
+
+pay_d03_i,82,237,3	script	Villager#ill_moon_1	1_F_ORIENT_01,6,6,{
+	end;
+
+OnTouch:
+	specialeffect EF_QUAKEBODY2;
+	npctalk "Ahhh... Ahhhhh!!";
+	initnpctimer;
+	disablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:970]:
```suggestion
	disablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 10]
```diff
@@ -0,0 +1,1039 @@
...
+	areamonster "pay_d03_i",81,236,83,238,"Resentful Sohee",3762,3;
+	end;
+
+OnTimer60000:
+	stopnpctimer;
+	enablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:976]:
```suggestion
	enablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 11]
```diff
@@ -0,0 +1,1039 @@
...
+pay_d03_i,254,206,3	script	Villager#ill_moon_2	4_F_01,6,6,{
+	end;
+
+OnTouch:
+	specialeffect EF_QUAKEBODY2;
+	npctalk "Something is not right!";
+	sleep2 2000;
+	npctalk "Stop it... It's coming, agh!!";
+	areamonster "pay_d03_i",252,204,255,208,"Resentful Munak",3760,3;
+	disablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:989]:
```suggestion
	disablenpc()
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 12]
```diff
@@ -0,0 +1,1039 @@
...
+	end;
+
+OnTimer60000:
+	stopnpctimer;
+	enablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:994]:
```suggestion
	enablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 13]
```diff
@@ -0,0 +1,1039 @@
...
+	mes "[ Villager ]";
+	mes "Are you here to subjugate us?!";
+	mes "We've done nothing wrong!";
+	mes "Can't you tell from the sacred creatures!";
+	next;
+	mes "[ Villager ]";
+	mes "Ah... The fox lady...";
+	close2;
+	initnpctimer;
+	disablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:1008]:
```suggestion
	disablenpc());
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 14]
```diff
@@ -0,0 +1,1039 @@
...
+	npctalk "H-help!! I don't want to die!";
+	areamonster "pay_d03_i",79,136,81,138,"Resentful Bongun",3761,3;
+	end;
+
+OnTimer60000:
+	stopnpctimer;
+	enablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:1015]:
```suggestion
	enablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 15]
```diff
@@ -0,0 +1,1039 @@
...
+	mes "You...";
+	mes "Who's that strange mage?";
+	mes "Why do you...?";
+	next;
+	mes "[ Villager ]";
+	mes "You! You're no better than them!";
+	mes "What have we done to you?";
+	close2;
+	initnpctimer;
+	disablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:1030]:
```suggestion
	disablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 16]
```diff
@@ -0,0 +1,1039 @@
...
+	npctalk "Arrgghh-!";
+	areamonster "pay_d03_i",158,258,162,262,"Resentful Bongun",3761,3;
+	end;
+
+OnTimer60000:
+	stopnpctimer;
+	enablenpc strnpcinfo(0);
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:1037]:
```suggestion
	enablenpc();
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 17]
```diff
@@ -0,0 +1,1039 @@
...
+		mes "We should grasp the situation first.","It'll be great if you can help us, Adventurer.";
+		close;
+	}
+	if (illusion_moonlight < 10) {
+		mes "[ Jagyeom ]";
+		mes "Please observe proper manners inside the palace.";
+		close;
+	}
+	mes "[ Jagyeom ]";
+	mes "The Moonlight Flower are recreating the past in their nightmares.","In a way, they're also victims.";
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:549]:
```suggestion
	mes "The Moonlight Flower are recreating the past in their nightmares.";
	mes "In a way, they're also victims.";
```

File: npc/re/merchants/enchan_illusion.txt [hunk 18]
```diff
@@ -0,0 +1,195 @@
...
+			mes "[ Gemcutter ]";
+			mes "As you may have guessed, your equipment will transform into something new after this.","In other words, ^0000FFIt'll lose its current refining levels, cards and enchantments.^000000";
+			next;
+			mes "[ Gemcutter ]";
+			mes "And I need ^0000FFIllusion Stones and some other materials^000000 to upgrade your equipment.","Pick an item you want and I'll tell you what I need.";
+			next;
+			for (.@i = 0; .@i < getarraysize(.weapons$); .@i++) {
+				explode(.@T$,.weapons$[.@i],"|");
+				.@item_id = atoi(.@T$[0]);
+				.@menu$ += getitemname(.@item_id) +":";
```
Comment (Atemo) [inline npc/re/merchants/enchan_illusion.txt:61]: Sometimes some items name contain ":", I suggest to always use `replacestr` to potentially remove ":"

File: npc/re/quests/quests_16_2_illusion.txt [hunk 19]
```diff
@@ -0,0 +1,1039 @@
...
+		next;
+		select("Have you seen someone named Muyeon?");
+		mes "[ Young Scholar ]";
+		mes "Muyeon? Do you mean Muyeon the Soul Linker?","Well, technically, she's still an apprentice, but that's not important.";
+		next;
+		mes "[ Young Scholar ]";
+		mes "I'm waiting for her right now.","She ran into the cave, as if she was possessed by something.";
+		next;
+		mes "[ Young Scholar ]";
+		mes "Junghee? I mean, a soldier followed her in, but neither of them came out yet.","I was debating if I should go after them.";
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:158]: Please make the sentence in 2 lines for this one and the others
```suggestion
		mes "Junghee? I mean, a soldier followed her in, but neither of them came out yet.";
		mes "I was debating if I should go after them.";
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 20]
```diff
@@ -0,0 +1,1039 @@
...
+		mes "[ Young Scholar ]";
+		mes "Muyeon must have gone to the ^0000CDDetached Payon Palace.^000000","I'm going there do you want to come with me?";
+		next;
+		illusion_moonlight = 2;
+		completequest 7776;
+		setquest 7777;
+        if (select("Go with him.:Go separately.") == 2) {
+			mes "[ Young Scholar ]";
+			mes "Muyeon must have gone to the ^0000CDDetached Payon Palace.^000000","Why don't you go talk to her?","I'll talk to the Elders.";
+			end;
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:205]: Please replace the end by close when the message box is active

File: npc/re/quests/quests_16_2_illusion.txt [hunk 21]
```diff
@@ -0,0 +1,1039 @@
...
+	if (checkquest(7784,HUNTING) == 2) {
+		mes "[ Junghee ]";
+		mes "It doesn't matter if they're real or not. They're just as threatening.","Good Job.";
+		next;
+		mes "[ Junghee ]";
+		mes "Chunghae's almost done investigating this place. I'd better get ready to leave.";
+		erasequest 7784;
+		setquest 7785;
+		getitem 25271,1;
+		questinfo_refresh();
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:820]: questinfo_refresh is useless here
```suggestion
```

File: npc/re/quests/quests_16_2_illusion.txt [hunk 22]
```diff
@@ -0,0 +1,1039 @@
...
+	if (checkquest(7786,HUNTING)  == 2) {
+		mes "[ Junghee ]";
+		mes "It doesn't matter if they're real or not.","I can symphatize with the soldier.";
+		next;
+		mes "[ Junghee ]";
+		mes "All he wanted to do was protect others, and he died without fulfilling that wish.","Thank you for helping them rest.";
+		erasequest 7786;
+		setquest 7787;
+		getitem 25271,1;
+		questinfo_refresh();
```
Comment (Atemo) [inline npc/re/quests/quests_16_2_illusion.txt:832]: questinfo_refresh is useless here
```suggestion
```

**Comment by secretdataz** [comment]
Please add missing skills for the MVP.
https://www.divine-pride.net/database/monster/3758/angry-moonlight-flower

**Comment by Everade** [comment]
> Please add missing skills for the MVP. https://www.divine-pride.net/database/monster/3758/angry-moonlight-flower

Which ones do you believe are missing?
I counted 30, and they should be all there.

File: npc/re/merchants/enchan_illusion_16_2.txt [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+			close;
+		}
+		setarray .@reward_id[0],
+			26109,		// Illusion Staff of Bordeaux
+			28725,		// Illusion Moonlight Dagger
+			16063,		// Illusion Long Mace
+			26007;		// Illusion Spectral Spear
+		.@size = getarraysize(.@reward_id);
+
+		.@refine_req = 9;
```
Comment (Atemo) [inline npc/re/merchants/enchan_illusion_16_2.txt:48]: The min refine is +7 in your copy, +9 in mine. It seems to be +9 according to several source, has it been changed in KRO ?

Comment (Everade) [reply npc/re/merchants/enchan_illusion_16_2.txt:48]: I've only verificated the item ids and prices, not the refine requirements. So could be an error.

**Review by Atemo — APPROVED** [review]

**Pull request merged**